src/main/scala/akka/projection/testing/LoadGeneration.scala

```diff
@@ -0,0 +1,108 @@
+
+  case class TestSummary(name: String, expectedMessages: Long)
+
+  def apply(settings: EventProcessorSettings, shardRegion: ActorRef[ShardingEnvelope[ConfigurablePersistentActor.Command]], source: DataSource): Behavior[RunTest] = Behaviors.setup { ctx =>
+    Behaviors.receiveMessage[RunTest] {
+      rt: RunTest =>
+        ctx.spawn(LoadTest(settings, rt.name, shardRegion, source), s"test-${rt.name}") ! Start(rt)
+        Behaviors.same
+    }
+  }
+
+}
+
+object LoadTest {
+
+  val threadSafeLog = LoggerFactory.getLogger("load-test")
+
+  sealed trait Command
+
+  case class Start(test: RunTest) extends Command
+
+  private case class StartValidation() extends Command
+
+  private case class LoadGenerationFailed(t: Throwable) extends Command
+
+  def apply(settings: EventProcessorSettings, testName: String, shardRegion: ActorRef[ShardingEnvelope[ConfigurablePersistentActor.Command]], source: DataSource): Behavior[Command] = Behaviors.setup { ctx =>
+    import akka.actor.typed.scaladsl.AskPattern._
+    // asks are retried
+    implicit val timeout: Timeout = 1.seconds
+    implicit val system: ActorSystem[Nothing] = ctx.system
+    implicit val ec: ExecutionContextExecutor = system.executionContext
+    implicit val scheduler: Scheduler = system.toClassic.scheduler
+
+    Behaviors.receiveMessagePartial[Command] {
+      case Start(RunTest(name, actors, eventsPerActor, replyTo, numberOfConcurrentActors, t)) =>
+        threadSafeLog.info("TestPhase: Starting load generation")
+        val expected: Int = actors * eventsPerActor
+        val total = expected * settings.nrProjections
+        replyTo ! TestSummary(name, expected * settings.nrProjections)
+        val startTime = System.nanoTime()
+
+        // The operation is idempotent so retries will not affect the final event count
+        val testRun: Source[StatusReply[Done], NotUsed] = Source(1 to actors)
+          .mapAsyncUnordered(numberOfConcurrentActors) { id =>
+            val pid = s"${testName}-$id"
+            val retried: Future[StatusReply[Done]] = retry(
+              () => {
+                threadSafeLog.info("Sending message to pid {}", pid)
+                shardRegion.ask[StatusReply[Done]] { replyTo => ShardingEnvelope(pid, ConfigurablePersistentActor.PersistAndAck(eventsPerActor, s"actor-$id-message", replyTo, testName))
+                }
+              }, 20, 1.second, 30.seconds, 0.1)
+
+
+            retried.transform {
+              case s@Success(_) => s
+              case Failure(_) => Failure(new RuntimeException(s"Load generation failed for persistence id ${pid}")) // this will be an ask timeout
+            }
+          }
+
+        ctx.pipeToSelf(testRun.run()) {
+          case Success(_) => StartValidation()
+          case Failure(t) => LoadGenerationFailed(t)
+        }
+
+        Behaviors.receiveMessagePartial[Command] {
+          case StartValidation() =>
```

Comment:
here it would be nice with logging the write throughput
```
val durationMs = math.max(1, (System.nanoTime() - startTime) / 1000 / 1000)
ctx.log.info("Write throughput [{}] events/s", 1000 * total / durationMs)
```

I noticed that it was much reduced when changing to the synchronous tag writing https://github.com/akka/akka-persistence-cassandra/pull/841.
Write throughput 2809 vs 11775

One could think that was because the tags were not fully written yet, but the total rate was also reduced.
Total Rough rate 2531 vs 4878

I used
```
curl  -X POST  -v --data '{"name":"","nrActors":20000, "messagesPerActor": 10, "concurrentActors": 200, "timeout": 60000}' --header "Content-Type: application/json" http://127.0.0.1:8051/test
```

and had changed to grouped projection and `nr-projections = 1` and `projection-failure-every = 1000000`

Reply:
So must be a mixture of the lack of batching and the waiting.

To investigate this further i'll see about getting metrics out to say how many writes to the database we do for the old impl and the sync one. I'd guess we do vastly more without batching which is causing the throughput drop rather than the future juggling but I could be wrong!

Reply:
and it might have been more equal if increasing concurrentActors (although I expect that the sync write will be much more stable and avoid overload, which was the primary purpose of this)